### PR TITLE
Type-safe Transformer Language Model

### PR DESCRIPTION
**Pull request by tscholak**
- [x] type-safe `chunk`
- [x] improved type-safe `matmul` on par with https://pytorch.org/docs/stable/torch.html#torch.matmul
- [x] type-safe `softmax`, improved type-save `logSoftmax`
- [x] type-safe `dropout`
- [ ] tbd

### CODE REVIEW

**Comment by tscholak** [comment]
@junjihashimoto I've made this into a proper PR, feel free to merge it in.

**Comment by junjihashimoto** [comment]
This is awesome work.
Thx!
